core/src/main/java/smile/validation/CrossValidation.java

```diff
@@ -137,11 +139,11 @@ public CrossValidation(int n, int k, boolean permutate) {
      * Runs cross validation tests.
      * @return the predictions.
      */
-    public int[] classification(Formula formula, DataFrame data, BiFunction<Formula, DataFrame, DataFrameClassifier> trainer) {
+    public int[] classification(Formula formula, DataFrame data, BiFunction<Formula, DataFrame, SoftClassifier<Tuple>> trainer) {
```

Comment:
It may not be a `SoftClassifier`.

Reply:
Yes I wanted to leave DataFrameClassifier in the first place, but I had some type errors because the DataFrameClassifier interface do not extend the Classifier interface and thus the classifiers attribute of the OneVersusRest cannot be a list of DataFrameClassifier s. In contrast, the SoftClassifier interface extends the Classifier interface so it is possible for the classifiers attribute of the OneVersusRest class to be an array of SoftClassifier s. Moreover, to the best of my knowledge, all classifiers that implements DataFrameClassifier also implements SoftClassifier, that is why I thought this would not be an issue. But I might be wrong on that last point.